### ISSUE TITLE
efficient way of saving models?

### ISSUE THREAD

**Issue by RaiAmanRai** [huggingface/setfit#219]
Hi guys, pretty interesting project.

I was wondering if there is any way to save the finetuned models.

So for example, if I finetuned a couple of say, `sentence-transformers/paraphrase-mpnet-base-v2` models, the major difference between them is just the weights of final few layers, weights for the rest of the model mostly remains the same, so is there a way to  efficiently save the necessary final few layers thus reducing the size of models, repetedly being saved.

This way one could save, the disk space by a lot.

And apart form that, while inferencing, I don't have to load multiple huge models and instead I could have just one model containg the common freezed layers that give me some common features and just has to host the final few layers with custom classes.

**Comment by tomaarsen** [comment]
Hey @RaiAmanRai,

I'm afraid that the weights of the sentence transformer, i.e. the first layers, do get modified significantly. To back this up, let's discuss the model briefly. It consists of a sentence transformer body and a classifier head, e.g. a logistic regression head. This model is trained in two subsequent steps:

1. The sentence transformer body is fully finetuned using contrastive learning.
2. Only then, the classifier head is fully trained, without changing the sentence transformer weights.

If we want to save disk space by keeping the original sentence transformer weights, then step 1 is discarded. 

Out of curiosity, I went and performed a quick experiment on this using the first example from the [README's Usage section](https://github.com/huggingface/setfit#usage). If I run that experiment normally, then I report an accuracy of 83.97% (σ 3.46% at sample size of 4). If I remove the sentence transformer finetuning, then the performance drops to a consistent 73.62%. This performance drop is what you may expect if you don't want to update the sentence transformer weights.

At that point, the approach just reduces to learning a classifier head on a sentence transformer. After all, the classifier head is not trained in a contrastive way, which (in my opinion) is the main strength of this work.

- Tom Aarsen

**Comment by RaiAmanRai** [comment]
Hi @tomaarsen 

Thanks for the answer. Now I get the idea how important the contrastive learning part is and couldn't compromise with the such a drop in accuracy.

Just out of curiosity, did you guys try the p-tuning aproach, which basically uses the concept of soft-tokens and gives better results compared to prompt engineering. Was wondering what would that look like in comparison to this approach.

Thanks

**Comment by tomaarsen** [comment]
Hey @RaiAmanRai 

Firstly, I want to clarify that I'm not an author of SetFit. I'm simply an enthusiast interested in the work. With other words, I can't say whether the authors looked into p-tuning. 

I had not read up on p-tuning (and p-tuning v2) as of just now, and my understanding is that it tries to learn which prompts are effective. However, SetFit dispenses with prompts altogether, as mentioned briefly in the paper & repo README. Thus, I suspect that p-tuning may not be (directly) applicable for SetFit. 
That said, I'm not well enough aware of p-tuning to be able to say that conclusively.

- Tom Aarsen
